File: tools/binding_compliance/conformance/families/config_yaml_values.py
"""Bounded raw YAML accessor coverage over complete observed projections."""

from collections.abc import Mapping

from ..coverage import CoveragePredicate

YAML_VALUE_KEYS = (
    "classic_version",
    "classic_version_date",
    "crashgen_name_field",
    "crashgen_latest_og",
    "warn_noplugins",
    "warn_outdated",
    "xse_acronym",
    "autoscan_text",
    "game_version",
    "game_root_name_field",
    "classic_game_hints",
    "classic_records_list",
    "crashgen_ignore_og",
    "game_ignore_plugins",
    "game_ignore_records",
    "ignore_list",
    "suspects_error_keys",
    "suspects_error_values",
    "suspects_stack_keys",
    "mods_core_keys",
    "mods_core_values",
    "mods_core_names",
    "mods_core_gpus",
    "mods_conf_mod_a",
    "mods_conf_mod_b",
    "mods_conf_name_a",
    "mods_conf_name_b",
    "mods_conf_descriptions",
    "mods_conf_fixes",
    "mods_conf_links",
    "mods_core_count",
    "mods_conf_count",
    "mods_freq_entries",
    "mods_solu_entries",
    "suspects_error_rules",
    "suspects_stack_rules_metadata",
    "mods_conf_has_fixes",
    "suspects_stack_count_rules_for_id",
    "get_crashgen_name",
    "get_game_root_name",
    "get_crashgen_ignore",
)


def matches_yaml_values(observation):
    """Require the complete public accessor projection and exact source envelope."""
    result = observation.get("result")
    values = result.get("yamlValues") if isinstance(result, Mapping) else None
    return (
        set(observation) == {"result", "error", "files"}
        and observation["error"] is None
        and isinstance(values, Mapping)
        and set(values) == set(YAML_VALUE_KEYS)
        and values.get("classic_version") == result.get("classicVersion")
        and values.get("get_crashgen_name") == values.get("crashgen_name_field")
        and values.get("get_game_root_name") == values.get("game_root_name_field")
    )


YAML_VALUES_PREDICATE = CoveragePredicate(
    "config-operations.yaml-values",
    "config-operations.yaml-values",
    "config-operations.load-explicit",
    "values",
    ("YamlDataCore",),
    matches_yaml_values,
    runtime_operations=(
        None,
        "createYamlDataFromContent",
        "YamlData.from_yaml_content",
        "YamlData.__repr__",
        "yaml_data_classic_version",
        "yaml_data_classic_version_date",
        "yaml_data_crashgen_name_field",
        "yaml_data_crashgen_latest_og",
        "yaml_data_warn_noplugins",
        "yaml_data_warn_outdated",
        "yaml_data_xse_acronym",
        "yaml_data_autoscan_text",
        "yaml_data_game_version",
        "yaml_data_game_root_name_field",
        "yaml_data_classic_game_hints",
        "yaml_data_classic_records_list",
        "yaml_data_crashgen_ignore_og",
        "yaml_data_game_ignore_plugins",
        "yaml_data_game_ignore_records",
        "yaml_data_ignore_list",
        "yaml_data_suspects_error_keys",
        "yaml_data_suspects_error_values",
        "yaml_data_suspects_stack_keys",
        "yaml_data_mods_core_keys",
        "yaml_data_mods_core_values",
        "yaml_data_mods_core_names",
        "yaml_data_mods_core_gpus",
        "yaml_data_mods_conf_mod_a",
        "yaml_data_mods_conf_mod_b",
        "yaml_data_mods_conf_name_a",
        "yaml_data_mods_conf_name_b",
        "yaml_data_mods_conf_descriptions",
        "yaml_data_mods_conf_fixes",
        "yaml_data_mods_conf_links",
        "yaml_data_mods_core_count",
        "yaml_data_mods_conf_count",
        "yaml_data_mods_freq_entries",
        "yaml_data_mods_solu_entries",
        "yaml_data_suspects_error_rules",
        "yaml_data_suspects_stack_rules_metadata",
        "yaml_data_mods_conf_has_fixes",
        "yaml_data_suspects_stack_count_rules_for_id",
        "yaml_data_get_crashgen_name",
        "yaml_data_get_game_root_name",
        "yaml_data_get_crashgen_ignore",
    ),
)

File: business-logic/classic-scanlog-core/tests/semantic_conformance/config_yaml_values.rs
//! Native YAML fields and empty structured collection observations.
use classic_config_core::YamlDataCore;
use serde_json::{Value, json};

/// Read raw and compatibility views of the same core-owned parsed data.
pub(super) fn observe(data: &YamlDataCore) -> Value {
    let mut values = serde_json::Map::new();
    values.insert("classic_version".into(), json!(data.classic_version));
    values.insert(
        "classic_version_date".into(),
        json!(data.classic_version_date),
    );
    values.insert("crashgen_name_field".into(), json!(data.crashgen_name));
    values.insert("crashgen_latest_og".into(), json!(data.crashgen_latest_og));
    values.insert("warn_noplugins".into(), json!(data.warn_noplugins));
    values.insert("warn_outdated".into(), json!(data.warn_outdated));
    values.insert("xse_acronym".into(), json!(data.xse_acronym));
    values.insert("autoscan_text".into(), json!(data.autoscan_text));
    values.insert("game_version".into(), json!(data.game_version));
    values.insert("game_root_name_field".into(), json!(data.game_root_name));
    values.insert("classic_game_hints".into(), json!(data.classic_game_hints));
    values.insert(
        "classic_records_list".into(),
        json!(data.classic_records_list),
    );
    values.insert("crashgen_ignore_og".into(), json!(data.crashgen_ignore));
    values.insert(
        "game_ignore_plugins".into(),
        json!(data.game_ignore_plugins),
    );
    values.insert(
        "game_ignore_records".into(),
        json!(data.game_ignore_records),
    );
    values.insert("ignore_list".into(), json!(data.ignore_list));
    values.insert(
        "suspects_error_keys".into(),
        json!(
            data.suspect_error_rules
                .iter()
                .map(|entry| entry.id.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "suspects_error_values".into(),
        json!(
            data.suspect_error_rules
                .iter()
                .map(|entry| entry.name.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "suspects_stack_keys".into(),
        json!(
            data.suspect_stack_rules
                .iter()
                .map(|entry| entry.id.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_core_keys".into(),
        json!(
            data.game_mods_core
                .iter()
                .map(|entry| entry.detect.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_core_values".into(),
        json!(
            data.game_mods_core
                .iter()
                .map(|entry| entry.description.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_core_names".into(),
        json!(
            data.game_mods_core
                .iter()
                .map(|entry| entry.name.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_core_gpus".into(),
        json!(
            data.game_mods_core
                .iter()
                .map(|entry| entry.gpu.clone().unwrap_or_default())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_mod_a".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.mod_a.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_mod_b".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.mod_b.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_name_a".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.name_a.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_name_b".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.name_b.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_descriptions".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.description.clone())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_fixes".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.fix.clone().unwrap_or_default())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "mods_conf_links".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.link.clone().unwrap_or_default())
                .collect::<Vec<_>>()
        ),
    );
    values.insert("mods_core_count".into(), json!(data.game_mods_core.len()));
    values.insert("mods_conf_count".into(), json!(data.game_mods_conf.len()));
    values.insert("mods_freq_entries".into(), json!(data.game_mods_freq.len()));
    values.insert("mods_solu_entries".into(), json!(data.game_mods_solu.len()));
    values.insert(
        "suspects_error_rules".into(),
        json!(data.suspect_error_rules.len()),
    );
    values.insert(
        "suspects_stack_rules_metadata".into(),
        json!(data.suspect_stack_rules.len()),
    );
    values.insert(
        "mods_conf_has_fixes".into(),
        json!(
            data.game_mods_conf
                .iter()
                .map(|entry| entry.fix.is_some())
                .collect::<Vec<_>>()
        ),
    );
    values.insert(
        "suspects_stack_count_rules_for_id".into(),
        json!(
            data.suspect_stack_rules
                .iter()
                .filter(|entry| entry.id == "absent-rule")
                .map(|entry| entry.stack_contains_at_least.len())
                .sum::<usize>()
        ),
    );
    values.insert("get_crashgen_name".into(), json!(data.get_crashgen_name()));
    values.insert(
        "get_game_root_name".into(),
        json!(data.get_game_root_name()),
    );
    values.insert(
        "get_crashgen_ignore".into(),
        json!(data.get_crashgen_ignore()),
    );
    Value::Object(values)
}
